packages/transaction-controller/src/TransactionController.ts

```diff
@@ -901,46 +923,55 @@ export class TransactionController extends BaseController<
   private async processApproval(
     transactionMeta: TransactionMeta,
     {
+      isExisting = false,
       requireApproval,
       shouldShowRequest = true,
     }: {
+      isExisting?: boolean;
       requireApproval?: boolean | undefined;
       shouldShowRequest?: boolean;
     },
   ): Promise<string> {
     const transactionId = transactionMeta.id;
     let resultCallbacks: AcceptResultCallbacks | undefined;
+    const { meta, isCompleted } = this.isTransactionCompleted(transactionId);
+    const finishedPromise = isCompleted
+      ? Promise.resolve(meta)
+      : this._waitForTransactionFinished(transactionId);
+
+    if (!isExisting && !isCompleted) {
+      try {
+        if (requireApproval !== false) {
+          const acceptResult = await this.requestApproval(transactionMeta, {
+            shouldShowRequest,
+          });
+          resultCallbacks = acceptResult.resultCallbacks;
+        }
 
-    try {
-      if (requireApproval !== false) {
-        const acceptResult = await this.requestApproval(transactionMeta, {
-          shouldShowRequest,
-        });
-        resultCallbacks = acceptResult.resultCallbacks;
-      }
-
-      const { meta, isCompleted } = this.isTransactionCompleted(transactionId);
-
-      if (meta && !isCompleted) {
-        await this.approveTransaction(transactionId);
-      }
-    } catch (error: any) {
-      const { meta, isCompleted } = this.isTransactionCompleted(transactionId);
-
-      if (meta && !isCompleted) {
-        if (error.code === errorCodes.provider.userRejectedRequest) {
-          this.cancelTransaction(transactionId);
+        const { isCompleted: isTxCompleted } =
+          this.isTransactionCompleted(transactionId);
 
-          throw ethErrors.provider.userRejectedRequest(
-            'User rejected the transaction',
-          );
-        } else {
-          this.failTransaction(meta, error);
+        if (!isTxCompleted) {
+          await this.approveTransaction(transactionId);
+        }
+      } catch (error: any) {
+        const { isCompleted: isTxCompleted } =
+          this.isTransactionCompleted(transactionId);
+        if (meta && !isTxCompleted) {
```

Comment:
If a `meta` reference is guaranteed now, should we remove this `meta` condition?

Reply:
It still can be `undefined` so to avoid `failTransaction` I can move the check to the first `if`.